#24652 -- SimpleTestCase should prevent execution of database queries

charettes opened the pull request:
- [x] Mention in the release notes.
- [x] Come up with a decent assertion message.

charettes commented:
Buildbot, test this please.

charettes commented:
I'm not sure it's worth documenting `allow_database_queries` since the documentation clearly states that:

> If you need any of the other more complex and heavyweight Django-specific features like:
>
> - Testing or using the ORM.
> - Database *fixtures*.
> - Test skipping based on database backend features.
> - The remaining specialized **assert*** methods.
>
> then you should use TransactionTestCase or TestCase instead.

And the assertion failure message raised is also quite explicit.

charettes commented:
Buildbot re-test this please.

timgraham commented on django/test/testcases.py:
```diff
@@ -777,6 +802,10 @@ class TransactionTestCase(SimpleTestCase):
     # This can be slow; this flag allows enabling on a per-case basis.
     serialized_rollback = False
 
+    # Since tests will be wrapped in a transactions if available or serialized
```
transaction
, or serialized if they are not available,

timgraham commented on docs/releases/1.9.txt:
```diff
@@ -505,6 +505,9 @@ Miscellaneous
 * The ``django.contrib.sites.models.Site.domain`` field was changed to be
   :attr:`~django.db.models.Field.unique`.
 
+* In order to enforce test isolation database queries are not allowed
```
isolation,

timgraham commented on django/test/testcases.py:
```diff
@@ -140,6 +140,19 @@ def message(self):
+
+    def __call__(self):
+        raise AssertionError(
+            "SimpleTestCase shouldn't be used to test features relying on the ORM. "
```
perhaps more generally, "Database queries aren't allowed in SimpleTestCase." (since queries can be executed in ways other than just the oRM)

timgraham commented on tests/test_utils/tests.py:
```diff
@@ -914,3 +914,21 @@ def test_override_staticfiles_dirs(self):
+
+
+class DisallowedDatabaseQueriesTests(SimpleTestCase):
+    def runTest(self):
```
Just curious if there is a reason for the `runTest` name instead of the usual non-camel case name.

charettes replied:
I think I chose this pattern because I wanted to explicitly state the classes shouldn't reused for other tests but since this pattern is not used anywhere else in the suite I'll stick to the commonly used one.

timgraham commented:
I would document that SimpleTestCase prevents database queries in topics/testing/tools.txt, but I wouldn't document `allow_database_queries` as you suggested. Please backport the isolation fixes to 1.8 if they merge cleanly. Thanks, this should be very helpful!

charettes commented:
Ok I adjusted the patch based on your review and added documentation in `topics/testing/tools.txt` about the database queries prevention behaviour of `SimpleTestCase`. Let me know if it makes sense.

I'll make sure to backport isolation fixes to `1.8.x`.

akaariai commented:
Documenting the flag could be a good idea. I have a feeling we have done enough "break user code in the name of safety" lately (assigning None to non-nullable foreign keys comes to mind here).

There are valid use cases where you want to use SimpleTestCase, but don't want to use either TransactionTestCase nor TestCase (the former causes database flush and performance issues, the latter disallows transactions).

If somebody is currently using SimpleTestCase and is happy about it, I don't believe we have reason to force everybody to update their code here. But telling them that there might be a problem in their test setup is definitely a good idea.

charettes commented:
Documented `allow_database_queries` based on Anssi's comments.

timgraham commented on django/test/testcases.py:
```diff
@@ -148,6 +161,10 @@ class SimpleTestCase(unittest.TestCase):
     _overridden_settings = None
     _modified_settings = None
 
+    # Tests shouldn't be allowed to mess with the database since
```
mess with -> query?

timgraham commented on docs/releases/1.9.txt:
```diff
@@ -505,6 +505,12 @@ Miscellaneous
 
+* In order to enforce test isolation, database queries are not allowed
+  by default in :class:`~django.test.SimpleTestCase` tests anymore. You
+  can disable this behavior by explicitly setting the
```
chop "explicitly"

timgraham commented on docs/topics/testing/tools.txt:
```diff
@@ -606,6 +606,16 @@ features like:
+
+    .. versionadded:: 1.9
+
+    :class:`~SimpleTestCase` will prevent database queries by raising
```
SimpleTestCase disallows database queries by default. This helps to avoid executing write queries which will affect other tests since each SimpleTestCase test isn't run under a transaction. If you aren't concerned about this problem, you can disable...

timgraham commented on docs/topics/testing/tools.txt:
```diff
@@ -606,6 +606,16 @@ features like:
+    :class:`~SimpleTestCase` will prevent database queries by raising
+    an assertion error since it doesn't enforce data isolation by reseting
+    database to known state between tests. You can disable this behavior by
+    explicitly setting the ``allow_database_queries`` flag to ``True`` on
```
chop "explicitly"
flag -> class attribute?
